crates/astria-telemetry/src/metrics/builders.rs

```diff
@@ -0,0 +1,277 @@
+    pub fn with_global_recorder(mut self, use_global_recorder: bool) -> Self {
+        self.use_global_recorder = use_global_recorder;
+        self
+    }
+
+    /// Registers the buckets and metrics as specified in `T::set_buckets` and `T::register`
+    /// respectively, starts the http server if enabled, sets the global metrics recorder if
+    /// requested and returns a new metrics object of type `T` along with a handle for rendering
+    /// current metrics.
+    // allow: no useful error info can be added without writing excessive details.
+    #[allow(clippy::missing_errors_doc)]
+    pub fn build<T: Metrics>(self, config: &T::Config) -> Result<(T, Handle), Error> {
+        // Apply settings to the prometheus builder.
+        let mut prometheus_builder = PrometheusBuilder::new();
+        if !self.service_name.is_empty() {
+            prometheus_builder = prometheus_builder.add_global_label("service", self.service_name);
+        }
+        if let Some(listening_address) = &self.listening_address {
+            let addr: SocketAddr = listening_address.parse()?;
+            prometheus_builder = prometheus_builder.with_http_listener(addr);
+        }
+
+        // Set the histogram buckets.
+        let mut bucket_builder = BucketBuilder {
+            builder: prometheus_builder,
+            buckets: HashMap::new(),
+        };
+        T::set_buckets(&mut bucket_builder, config)?;
+        let histograms_with_buckets = bucket_builder.histogram_names();
+
+        // Consume the prometheus builder, yielding a recorder and a future for running the exporter
+        // server (this will be a no-op if the server isn't configured to run).
+        let (recorder, exporter_fut) = if self.listening_address.is_some() {
+            bucket_builder
+                .builder
+                .build()
+                .map_err(|error| Error::StartListening(error.into()))?
+        } else {
+            let recorder = bucket_builder.builder.build_recorder();
+            let fut: ExporterFuture = Box::pin(async move { Ok(()) });
+            (recorder, fut)
+        };
+        let handle = Handle::new(recorder.handle());
+
+        // Register individual metrics.
+        let mut registering_builder = RegisteringBuilder::new(recorder);
+        let metrics = T::register(&mut registering_builder, config)?;
+
+        // Ensure no histogram buckets were left unassigned.
+        let unassigned: HashSet<_> = histograms_with_buckets
+            .difference(&registering_builder.histograms)
+            .cloned()
+            .collect();
+        if !unassigned.is_empty() {
+            return Err(Error::BucketsNotAssigned(unassigned));
+        }
+
+        // Run the exporter server and set the global recorder if requested.
+        tokio::spawn(exporter_fut);
+        if self.use_global_recorder {
+            metrics::set_global_recorder(registering_builder.recorder)
+                .map_err(|_| Error::GlobalMetricsRecorderAlreadySet)?;
```

Comment:
Does this need a https://doc.rust-lang.org/stable/std/sync/struct.LazyLock.html similar to what we do when setting up tracing in tests?

I don't know how metrics behaves if one tries to register multiple global recorders - tracing panics. Maybe we can provide an extra flag `use_test_protection` (or something less asinine) that ensures that no more than 1 global recorder is set?

(In services it's IMO desirable to just panic because one should not have more than 1 global recorder).

Reply:
It does return an error if we try to register a global recorder more than once.  The error is being returned from from `astria_telemetry::Config::try_init` in the same way as for a tracing init error, so we're covered that way for production code.

In all our tests (where we use the `LazyLock` for the telemetry config) we're not setting the metrics up as part of the `astria_telemetry::configure()` flow, since we want to avoid using the global recorder.  So in all tests where your question applies, we're doing something like
```rust
let (metrics, metrics_handle) = metrics::ConfigBuilder::new()
    .set_global_recorder(false)
    .build(&())
    .unwrap();
```
meaning we won't have an error returned since we're not setting the global recorder.